specification/hugr.md

```diff
@@ -768,39 +990,62 @@ Ultimately though, we cannot avoid the "stringly" type problem if we
-
-To strike a balance then, every extension provides YAML that declares its opaque
-types and a number of named **OpDef**s (operation-definitions), which may be
-polymorphic in type. Each OpDef specifies one of two methods for how the type
-of individual operations is computed:
-
-1. A type scheme is included in the YAML, to be processed by a "type scheme interpreter"
-   that is built into tools that process the HUGR.
-2. The extension self-registers binary code (e.g. a Rust trait) providing a function
-   `compute_signature` that computes the type.
-
-Each OpDef may declare named type parameters---if so then the individual operation nodes
-in a HUGR will provide for each a static-constant "type argument": a value that in many
-cases will be a type. These type arguments are processed by the type scheme interpreter
-or the `compute_signature` implementation to compute the type of that operation node.
+### Extension Implementation
+
+To strike a balance then, every extension provides YAML (or equivalent Rust structs)
+that declares named opaque **TypeDef**s and **OpDef**s. These are (potentially-polymorphic)
+definitions of types and operations, respectively---polymorphism arises because both may
+declare any number TypeParams (as per [Type System](#type-system)). To use a TypeDef as a type,
+it must be instantiated with TypeArgs appropriate for its TypeParams, and similarly
+to use an OpDef as a node operation: each node stores a static-constant list of TypeArgs.
+
+For TypeDef's, any set of TypeArgs conforming to its TypeParams, produces a valid type.
+However, for OpDef's, greater flexibility is allowed: each OpDef *either*
+
+1. Provides a `Function` type, as per [Type System](#type-system), which may declare TypeParams; *or*
+2. The extension may self-register binary code (e.g. a Rust trait) providing a function
+   `compute_signature` that fallibly computes a `Function` type given some type arguments.
+   The operation declares the arguments required by the `compute_signature` function as a list
+   of TypeParams; if this successfully returns a `Function` type, then that may then require
+   additional TypeParams.
+
+For example, the TypeDef for `array` in the prelude declares two TypeParams: a `BoundedUSize`
+(the array length) and a `Type`. Any valid instantiation (e.g. `array<5, usize>`) is a type.
+Much the same applies for OpDef's that provide a `Function` type.
+
+However, for OpDefs providing binary `compute_signature`,
+* each node will provide TypeArgs for *both* the binary function *and* the `Function` type that returns
+* the operation is only valid if `compute_signature` succeeds, and returns a `Function` type
+  into whose TypeParams the *remaining* TypeArgs (in the node) fit. Note that this means
+  the binary function may use the values of its TypeArgs---specifically including `List`, `Tuple` and
+  `Opaque`---to determine the structure of the returned `Function` type such that the latter's own TypeParams
```

Comment:
```suggestion
  the binary function may use the values of its TypeArgs--specifically including `List`, `Tuple` and
  `Opaque`--to determine the structure of the returned `Function` type such that the latter's own TypeParams
```

Reply:
?? Searching Google for "parenthetic dashes" confirms that these should be em dashes, which is `---` ?